larpandoracontent/LArMonitoring/EventClusterValidationAlgorithm.cc

```diff
@@ -0,0 +1,337 @@
+            else if (valType == ValidationType::TRACK)
+                branchPrefix = "track_";
+            SetBranches(metrics, adjustedRandI, branchPrefix);
+        }
+
+        PANDORA_MONITORING_API(SetTreeVariable(this->GetPandora(), m_treeName, "event", m_eventNumber - 1));
+        PANDORA_MONITORING_API(SetTreeVariable(this->GetPandora(), m_treeName, "view", static_cast<int>(view)));
+        PANDORA_MONITORING_API(FillTree(this->GetPandora(), m_treeName));
+    }
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void EventClusterValidationAlgorithm::GetHitParents(const CaloHitList &caloHits,
+                                                    const ClusterList &clusters,
+                                                    std::map<const CaloHit *const, CaloHitParents> &hitParents) const
+{
+    for (const CaloHit *const pCaloHit : caloHits)
+    {
+        const MCParticle *pMainMC {nullptr};
+        const MCParticleWeightMap &weightMap {pCaloHit->GetMCParticleWeightMap()};
+        float maxWeight {std::numeric_limits<float>::lowest()};
+        for (const auto &[pMC, weight] : weightMap)
+        {
+            if (weight > maxWeight)
+                pMainMC = pMC;
+        }
+        if (pMainMC)
+        {
+            hitParents[pCaloHit] = CaloHitParents();
+            hitParents[pCaloHit].m_pMainMC = pMainMC;
+        }
+    }
+
+    for (const Cluster *const pCluster : clusters)
+    {
+        const CaloHitList &isolatedHits {pCluster->GetIsolatedCaloHitList()};
+        CaloHitList clusterCaloHits;
+        pCluster->GetOrderedCaloHitList().FillCaloHitList(clusterCaloHits);
+        clusterCaloHits.insert(clusterCaloHits.end(), isolatedHits.begin(), isolatedHits.end());
+        for (const CaloHit *const pCaloHit : clusterCaloHits)
+        {
+            // Ignoring the calo hit if truth matching is missing
+            if (hitParents.find(pCaloHit) == hitParents.end())
+                continue;
+            hitParents[pCaloHit].m_pCluster = pCluster;
+        }
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void EventClusterValidationAlgorithm::ApplyMCParticleMinSumHits(std::map<const CaloHit *const, CaloHitParents> &hitParents) const
+{
+    std::map<const MCParticle *const, int> mcSumHits;
+    for (const auto &[pCaloHit, parents] : hitParents)
+    {
+        const MCParticle *const pMainMC {parents.m_pMainMC};
+        if (mcSumHits.find(pMainMC) == mcSumHits.end())
+            mcSumHits[pMainMC] = 0;
+        mcSumHits.at(pMainMC)++;
+    }
+
+    for (auto it = hitParents.begin(), itNext = it; it != hitParents.end(); it = itNext)
+    {
+      itNext++;
+      if (mcSumHits.at(it->second.m_pMainMC) < m_minMCHitsPerView)
+        hitParents.erase(it);
+    }
+}
```

Comment:
I thnk my preference here would be for

```
    for (auto it = hitParents.begin(), it != hitParents.end();)
    {
        if (mcSumHits.at(it->second.m_pMainMC) < m_minMCHitsPerView)
            it = hitParents.erase(it);
        else
            ++it;
    }
```

Alternatively, I think this conditional probably allows you to do this with the std::erase function and a lambda function and avoid the for loop altogether.

Reply:
Yeah this method is cleaner, hadnt realised `erase` returns the next element. Could use `erase_if` for maps then could do this with a lambda but that is C++20 so maybe safer to leave it